[WiP] Skins With Dynamic Decks/Waveforms/Mixer that change with controller

MythicPalette opened the pull request:
These skins have been customized to handle a variable that a controller can use to swap visible decks with the deck toggle button.
https://www.youtube.com/watch?v=zVocq6VQ6U4
This along with updates to the controller mappings can add a new functionality to Mixxx fairly easily.

MythicPalette committed:
Initial Commit

These skins have been customized to handle a variable that a controller can use to swap visible decks with the deck toggle button.

Holzhaus commented:
Thanks, I think this is a cool feature that allows using 4 decks even on small screens. Could we integrate it into the actual Deere skin instead of adding a copy? If 2-deck mode is active, we could switch the decks depending on the CO, and do nothing in 4-deck mode.

ronso0 commented:
@LadyDefile 
Unfortunately this branch is based on a rather old state, before the LateNight redesign.
Can you try to rebase your work on top of current `main`?

ronso0 commented:
> Thanks, I think this is a cool feature that allows using 4 decks even on small screens. Could we integrate it into the actual Deere skin instead of adding a copy? If 2-deck mode is active, we could switch the decks depending on the CO, and do nothing in 4-deck mode. @ronso0 Do you think that's feasible or would that make skin maintenance very hard?

If we get this to work I think it should be in the regular skin, yes.
I'll have a look later on.

LadyDefile commented:
@Holzhaus My plan was more to get the ball rolling on this as an interface to be implemented as the skin designer wants. I figured if there was a "standardized method" for creating this effect it could be implemented in any skin. The way I recommend it is to  have a variable that tracks each side of the decks. In that case, if you create a dynamic 2-deck mode you could simply set it so that if deck1_active is true, show deck 1, if not, don't show any deck on that side or any other kind of implementation. 

@ronso0 I can absolutely do that. I didn't realize I had an old version of the LateNight skin. I can absolutely redo the skin based off the latest version. It's actually fairly easy to implement into skins. Honestly, the hardest part is figuring out where to put the settings in most cases, lol. I'll let you know how it goes with the reworked LateNight.

ronso0 commented:
Thanks for getting this started!
I already sketched a similiar concept earlier, and I consider the 'Compact Decks' recently introduced in LateNight step stone to a controller-focused solution.

First, let's me describe the scenario in which _I think_ a feature like this would be needed and why I think this implementation here wouldn't work IMO:
If someone plays with decks 1 and deck 4 only (decks 2 & 3 are idle) this doesn't matter and the current implementation works (but GUI decks need deck numbers to be sure GUI and controller are 'on the same page').

But my experience when playing with decks 1, 2 and 3 or 4 decks is that I need at least the basic info about ALL playing decks, especially when the controller doesn't give feedback on the deck it's currently NOT controlling.
Here however, 'hidden' decks incl. waveforms are entire inaccessible while it should be the other way around: allow deck control via the GUI when it's currently NOT accessible via controller. Same applies to the GUI mixer when playing with a 2-deck controller with deck switches: the mixer columns are swapped, too, but it doesn't make sense to have the current deck mixer controls both in the GUI and on the controller.

We should make some mockups how regular and minimal decks can be combined for a 4-deck controller scenario.

uklotzde commented:
Duplicating a whole skin is the wrong approach and not maintainable. We should instead try to extend the existing skins as @Holzhaus and @ronso already suggested.

Converting this PR into draft state.

MythicPalette committed:
Rebuilt LateNight Dynamic

LateNight Dynamic skin rebuilt from the newer version of the skin to support dynamic elements.

Removed LateNight (Dynamic) Since Ronso0 Mentioned making it part of the actual skin.

MythicPalette committed:
Deere Skin updated

Migrated Deere (Dynamic) from duplicate skin to extended skin

LadyDefile commented:
@uklotzde Changing from a duplicate skin to extended skin is easy enough, I just didn't want to suggest overwriting anyone else's work until such a time that I knew they were alright with it. Updated both of them to replace the original files with the new functionality.

@ronso0 How's that look to you?

MythicPalette committed:
Deere update

Deere is now the latest version and supports dynamic elements.

LadyDefile commented:
@uklotzde So I just went to go update the Deere (64 Samplers) and now I see what you mean by extension, lol. Now should I do an extension or no? My code doesn't remove any functionality from the original skins.

MythicPalette committed:
Extension conversion

- Reverted Deere to unedited version.
- Create a Deere (Dynamic) extension version.
- Reverted LateNight to unedited version.
- Updated LateNight paths (Added ../LateNight/ to the start of all paths. Without this, extensibility requires duplicating a lot more files than required.)
- Created LateNight (Dynamic) extended version.

LadyDefile commented:
@uklotzde How is that for an extension setup?

@ronso0 I had to update LateNight to support the extension model. It seems that none of the paths started with "../LateNight/" so it was either update the LateNight skin with the correct pathing or duplicate 80% of the original skin.

Be-ing commented:
Prior [Zulip discussion](https://mixxx.zulipchat.com/#narrow/stream/109171-development/search/.22active.20deck.22) on this topic

Be-ing commented:
I agree with @ronso0. What we should add is a visual indication of the "active" decks, not hiding the inactive ones.

Be-ing commented:
Thanks for getting started on this.

> I just didn't want to suggest overwriting anyone else's work until such a time that I knew they were alright with it.

Go ahead and change the files. Duplicating skins with a modification is impractical to review. Do not worry about changing someone else's work; this is why we discuss changes with pull requests before merging them.

We should not have special "controller" skins that add a new feature; all skins should support the same features. The 64 sampler skins are only intended as a quick hack around the issue of them requiring extending Mixxx's already long startup time. I don't think @uklotzde meant reusing that skin extension system for this purpose.

LadyDefile commented:
@Be-ing It's not so much that it's a special "controller" skin. Anything can be mapped to have the same effect which is why I took the approach I did. All that is required is that whatever you use works with Mixxx. The idea is that "any" device should be able to do this effect if setup to. All that's required for the skin to change is that your device's script has something to this effect: 

![image](https://user-images.githubusercontent.com/67084868/98431607-943d2200-207c-11eb-98d5-9a8129e9e679.png)

It's not that it's a "controller skin" but that it's a dynamic skin that can any device can be mapped to. I tested it with a phone app last night (DJ Pad). Used the midi controls to do the same thing as the controller.

Be-ing commented:
What hardware is being used with Mixxx is beside the point for that. If we add this feature, there's no reason these should be special skins apart from the usual skins.

LadyDefile commented:
@Be-ing So you think I should just make the changes part of the original skin then. Shouldn't be hard to do that honestly. Take maybe 10 minutes to do that if that's the way I should go about it.

LadyDefile commented:
Oh and @ronso0 The thing about the skin design I have is that it's up to the user. What I mean by this is that there is a setting for each waveform, mixer, and decks. So if they *just* want the decks to switch but keep the 4 waveforms and 4 mixer channels visible, they can do that. If they just want to set the mixer to show only the two active channels but keep the 4 decks and 4 waveforms, they can. Any mixture of those three settings. Examples:
Fully dynamic mode: 
![image](https://user-images.githubusercontent.com/67084868/98435538-61eded80-2099-11eb-9c36-c76e2d65db52.png)

Just dynamic decks & mixer:

![image](https://user-images.githubusercontent.com/67084868/98435557-9497e600-2099-11eb-8aa2-9cf55909e927.png)

Just dynamic waveforms:

![image](https://user-images.githubusercontent.com/67084868/98435568-aa0d1000-2099-11eb-8d81-6ffaec3d92e3.png)


As you can see, it adds the capability but doesn't "force" it on anyone. It's simply a setting on the right side that "can" be used. It's one that personally find helpful on smaller screens with no room for 4 decks,  and 4 waveforms.

Also, I think it's important that point out the long-term aim of this since it seems the focus is on the skins themselves. The goal is to get this sort of standardized and the bigger target here is the controller scripts themselves. With the controller scripts engaging with this, then it's an accessible feature that /can/ be used by anyone. A great example, a lot of people often talk about using Raspberri Pi but the screen they often use is extremely small, this is geat for something like that. Not to mention, the custom skin I use personally... Try to picture this with all four decks on screen at once:
![image](https://user-images.githubusercontent.com/67084868/98435698-c2c9f580-209a-11eb-8a6c-9ef845867171.png)

Thing is, sharing this skin right now is a completely mute point because I'm literally the only one that can use the dynamic deck feature of it at the moment considering nobody else has the custom mapping for it so really, my goal is just to have this "on offer" to the end-user but never forced on anyone.

ronso0 commented:
I'd like to go back to the drawing board here:
* can we make use of the existing minimal decks instead of completely hiding non-controllable decks?
* since we focus on a controller use case: do we need 'dynamic mixer columns'?
* should we toggle waveforms or keep them visbible under all circumstances? except the user hides them on purpose

Be-ing commented:
> a setting for each waveform, mixer, and decks

My first impression is that this would be overcomplicated and create an incoherent user experience. While it would be nice for a few users to have such detailed configuration options, I think for the vast majority of users it would be overwhelming. Should every controller script decide which parts of the GUI to toggle? What if someone uses a controller with a 7 inch screen and someone else uses the same controller with a 17 inch screen?

Let's take a step back. What exactly are we trying to accomplish here?

LadyDefile commented:
@ronso0 That would actual be super possible as well if I understand what you mean. Instead of hiding the deck completely, turn it into the mini version when inactive? That would be really, really easy to do actually. I could have that working in just a little while. 

The dynamic mixer columns I don't feel are necessary but I'm the type to lean to the "Have it and not need it" vs "Need it and not have it". The fact that you can disable the dynamic mixer makes it a "is there really a reason to remove the option?" situation for me. I'm a fan of highly customizable interfaces. 

As for the waveforms, they fit the same basic idea; someone, somewhere might use it and anyone that doesn't want to use it just doesn't turn it on. Keep in mind that *none* of the dynamic elements are on by default. By default, these skins behave exactly like before, the user has to knowingly enable the features. In this case even if you tried these skins knowing that these skins are customized like this, if you had the dynamic elements disabled would not be able to tell the difference between them and the original. They are exactly like the original skins but with an extra feature that you can use or ignore as you like. The dynamic elements aren't forced on any user.

@Be-ing It's 3 settings. That's a list. There is a setting that toggles dynamic waveforms on/off. There is a setting that turns dynamic decks on/off. There is a setting that toggles dynamic decks on/off. I think you misread that as each individual waveform, each individual mixer channel, etc. 

The controller script shouldn't decide *what* the skin does. I'll explain the functional process. There are two things at play here and it's important for both to be understood.

First is the controller. All the controller has to do is set the variables (variable group is [Controller], variables are deck1_active, deck2_active, deck3_active, deck4_active). That is all the controller does. If the skin doesn't have the variable then this doesn't cause any issues, I've tested this against default skins. So I repeat, the controller does one thing; i.e.:
`engine.setValue("[Controller]", "deck1_active", 0)` 

Now onto the skin. The way options work in the skin is fairly straight forward. You have a WidgetGroup that has things inside of it. You connect the "visible" property of that WidgetGroup to a variable. When the variable is 0, the object "disappears". When it's not 0, the object appears. This is where my skins come in. When dynamic is mode is enabled if [Controller],deck1_active is set to 0, deck 1 hides and deck 3 appears. If [Controller],deck2_active is set to 0, deck 2 hides and deck 4 appears. And if either is set to 1 it goes the other way (i.e. [Controller],deck1_active set to 1, deck 1 becomes visible). Keep in mind though, if the user has "Dynamic Decks" turned off, then the decks don't change. Same for "Dynamic Waveforms", if that is disabled, it just simply shows all 4 waveforms. If "Dynamic Mixer" is disabled, it just shows a 4 channel mixer.

Now the last thing you asked, "What are you trying to accomplish here?"

My goal has two steps:
- First, give users the *option* to use a dynamic interface (like Serato and Virtual DJ offer).
- Get controller mappings to set the deckX_active variables when they are toggled. (Comes later)

I'm going to upload a YouTube video that more accurately demonstrates *what* these skins do. I'll update you when that is ready.

LadyDefile commented:
Actually, I have this video that shows the options in use https://www.youtube.com/watch?v=zVocq6VQ6U4

As you can see, you enable each section manually and at the start of the video you'll notice, it's working in normal 4 deck mode.
